### PR TITLE
Feature/0180 choice uniqueness

### LINKED ISSUE #345: [SDL 0180] Broaden Choice Uniqueness
Proposal: [Broaden Choice Uniqueness](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0180-broaden-choice-uniqueness.md)

>This proposal broadens choice and sub-menu uniqueness to allow `PerformInteraction` and sub-menu `AddCommand` primary text to be identical.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/530

Steering Committee Decision:

> The Steering Committee voted to accept this proposal with the following revisions: numbers should be added after duplicate(s) for older versions of core, and documentation should call out that while the primary text doesn't have to be unique, it has to be clear.
> The proposal was revised to incorporate these revisions on July 12, 2018.

### PR DESCRIPTION
**Pull request by crokita**
Fixes #345 

NOTE: This PR relies on the choice set manager, so this PR should be merged first: #371 

### Risk
This PR makes no API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
Updated tests to account for text uniqueness changes, as well as adding a new test for the new choicesetmanager method.

#### Core Tests
Tested against current version of Manticore (core 7.0.0) 

### Summary
Allows for text duplication in the choice set manager, and has a fallback method for deduplication in older versions of SDL Core.

### CODE REVIEW

File: tests/managers/screen/choiceset/ChoiceSetManagerTests.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+            // Queued operation should be canceled
+            csm.dismissKeyboard(testCancelId);
+            Validator.assertEquals(callback.called, false);
+            Validator.assertEquals(callback2.called, true);
+            stub.restore();
+            stub2.restore();
+        });
+
+        it('testAddUniqueNamesToCells', function () {
```
Comment (renonick87) [inline tests/managers/screen/choiceset/ChoiceSetManagerTests.js:185]: This method has some linter errors relating to double quotes for strings, trailing spaces, and padding from blank lines.

Comment (crokita) [reply tests/managers/screen/choiceset/ChoiceSetManagerTests.js:185]: Addressed

**Review by renonick87 — CHANGES_REQUESTED** [review]
Will want to pull in the develop branch to fix the issues showing up from the linter.
The VideoStreaming PR is also present in this branch. It's fine if that gets merged first, but if not then it'll need to be reverted and squashed on this PR.
This may be unrelated to this PR but I've tried presenting a ChoiceSet with this code snippet
```
const choice1 = new ChoiceCell('cell')._setChoiceId(1);
const choice2 = new ChoiceCell('cell')._setChoiceId(2);
const choice3 = new ChoiceCell('cell')._setChoiceId(3);
const listener = new SDL.manager.screen.choiceset.ChoiceSetSelectionListener()
    .setOnError((error) => {
        console.log(`error: ${error}`);
    })
    .setOnChoiceSelected((choiceCell, triggerSource, rowIndex) => {
        console.log(`choiceCell: ${choiceCell}, triggerSource: ${triggerSource}, rowIndex: ${rowIndex}`);
    });
const choiceset = new ChoiceSet('choices', [choice1, choice2, choice3], listener);
screenManager.presentChoiceSet(choiceset, SDL.rpc.enums.InteractionMode.MANUAL_ONLY);
```
I get an error that says `INVALID_ID`. This happens regardless of whether I manually set the choiceID. I checked the PerformInteraction request and it looks like the ChoiceCells have all had their IDs set to 1 which could be what is causing the request to fail.

**Comment by crokita** [comment]
I have recreated the branch to exclude those other PR commits, accounting for the squashed changes of the newly merged choice set manager as well.

The reason for the choice set failing in the snippet above is because of the choice cells with the same names. The choice set manager will consider them as duplicate cell names and will be unable to use them to send a perform interaction to core (this will happen when testing on Manticore).

For this snippet to work you will need the build of core with the broaden choice uniqueness features so that duplicate primary texts are allowed

**Comment by crokita** [comment]
I see what you mean by the issue you found. Looks like the issue should be caught before the point where the RPC is attempted to be sent. See Joel's review in this PR for the ChoiceSet class: https://github.com/smartdevicelink/sdl_ios/pull/1886

**Comment by crokita** [comment]
Updated to have better behavior when sending exact duplicate choice cells, which should cause an error regardless of version used.

**Review by renonick87 — APPROVED** [review]

**Review by renonick87 — APPROVED** [review]

**Pull request merged**